src/runtime-rs/crates/hypervisor/ch-config/src/convert.rs

```diff
@@ -17,121 +17,214 @@ use std::convert::TryFrom;
 
 const DEFAULT_CH_MAX_PHYS_BITS: u8 = 46;
 
+const DEFAULT_VSOCK_CID: u64 = 3;
+
 impl TryFrom<NamedHypervisorConfig> for VmConfig {
-    type Error = anyhow::Error;
+    type Error = VmConfigError;
 
     fn try_from(n: NamedHypervisorConfig) -> Result<Self, Self::Error> {
-        let kernel_params = n.kernel_params;
+        let kernel_params = if n.kernel_params.is_empty() {
+            None
+        } else {
+            Some(n.kernel_params)
+        };
+
         let cfg = n.cfg;
-        let vsock_socket_path = n.vsock_socket_path;
-        let sandbox_path = n.sandbox_path;
+
+        let debug = cfg.debug_info.enable_debug;
+        let confidential_guest = cfg.security_info.confidential_guest;
+
+        let tdx_enabled = n.tdx_enabled;
+
+        let vsock_socket_path = if n.vsock_socket_path.is_empty() {
+            return Err(VmConfigError::EmptyVsockSocketPath);
+        } else {
+            n.vsock_socket_path
+        };
+
+        let sandbox_path = if n.sandbox_path.is_empty() {
+            return Err(VmConfigError::EmptySandboxPath);
+        } else {
+            n.sandbox_path
+        };
+
         let fs = n.shared_fs_devices;
 
-        let cpus = CpusConfig::try_from(cfg.cpu_info)?;
+        let cpus = CpusConfig::try_from(cfg.cpu_info).map_err(VmConfigError::CPUError)?;
 
-        let rng = RngConfig::try_from(cfg.machine_info)?;
+        let rng = RngConfig::from(cfg.machine_info);
 
         // Note how CH handles the different image types:
         //
-        // - An image is specified in PmemConfig.
+        // - A standard image is specified in PmemConfig.
         // - An initrd/initramfs is specified in PayloadConfig.
+        // - A confidential guest image is specified by a DiskConfig.
+        //   - If TDX is enabled, the firmware (`td-shim` [1]) must be
+        //     specified in PayloadConfig.
+        // - A confidential guest initrd is specified by a PayloadConfig with
+        //   firmware.
+        //
+        // [1] - https://github.com/confidential-containers/td-shim
         let boot_info = cfg.boot_info;
 
         let use_initrd = !boot_info.initrd.is_empty();
         let use_image = !boot_info.image.is_empty();
 
         if use_initrd && use_image {
-            return Err(anyhow!("cannot specify image and initrd"));
+            return Err(VmConfigError::MultipleBootFiles);
         }
 
         if !use_initrd && !use_image {
-            return Err(anyhow!("missing boot file (no image or initrd)"));
+            return Err(VmConfigError::NoBootFile);
         }
 
-        let initrd = if use_initrd {
-            Some(PathBuf::from(boot_info.initrd.clone()))
-        } else {
-            None
-        };
-
-        let pmem = if use_initrd {
+        let pmem = if use_initrd || confidential_guest {
             None
         } else {
-            let pmem = PmemConfig::try_from(&boot_info)?;
+            let pmem = PmemConfig::try_from(&boot_info).map_err(VmConfigError::PmemError)?;
```

Comment:
I wonder if it would not be better to use a `.context()` here instead, to avoid losing the original error.

Reply:
@c3d - This is a bit subtle but it's not losing the original error, it's returning it: `PmemConfig::try_from()` returns a `PmemConfigError` and the `map_err()` is returning that original error by wrapping it in a `VmConfigError::PmemError`, which is defined as:

```rust
#[error("Pmem config error: {0}")]
PmemError(PmemConfigError),
```